lib/evaluate/fold.cc

```diff
@@ -209,6 +212,102 @@ Expr<T> FoldOperation(FoldingContext &context, Designator<T> &&designator) {
       std::move(designator.u));
 }
 
+// Array constructor folding
+
+Expr<ImpliedDoIndex::Result> FoldOperation(
+    FoldingContext &context, ImpliedDoIndex &&iDo) {
+  auto iter{context.impliedDos.find(iDo.name)};
+  CHECK(iter != context.impliedDos.end());
+  return Expr<ImpliedDoIndex::Result>{iter->second};
+}
+
+template<typename T> class ArrayConstructorFolder {
+public:
+  explicit ArrayConstructorFolder(const FoldingContext &c) : context_{c} {}
+
+  Expr<T> FoldArray(ArrayConstructor<T> &&array) {
+    if (FoldArray(array.values)) {
+      std::int64_t n = elements_.size();
+      Expr<T> result{
+          Constant<T>{std::move(elements_), std::vector<std::int64_t>{n}}};
+      return result;
+    } else {
+      return Expr<T>{std::move(array)};
+    }
+  }
+
+private:
+  bool FoldArray(const CopyableIndirection<Expr<T>> &expr) {
+    Expr<T> folded{Fold(context_, Expr<T>{*expr})};
+    if (auto *c{UnwrapExpr<Constant<T>>(folded)}) {
+      // Copy elements in Fortran array element order
```

Comment:
Edit: From our discussion, I understand it is to avoid exposing the iterator and tempting people to calculate offsets on their own rather than using At which could lead to many bugs.
That is fine for me. I will try to centralize as much as possible this access pattern while folding elemental intrinsic to avoid duplicating it.